### PR TITLE
Implement computing normal and incident angle field

### PR DESCRIPTION
**Pull request by prybicki**

### CODE REVIEW

File: test/src/scene/incidentAngleTest.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+TEST_P(IncidentAngleTest, wall)
+{
+	// Parameters
+	auto&& [sensorPos, wallPosOffset, sensorRotDeg, wallRotDeg] = GetParam();
+	Vec3f sensorRotRad = (std::numbers::pi_v<float> / 180.0f) * sensorRotDeg;
+	Vec3f wallRotRad = wallRotDeg * Vec3f(std::numbers::pi_v<float> / 180.0f);
+
+	// Expectations
+	Vec3f expectedNormal = Mat3x4f::rotationRad(wallRotRad) * Vec3f{0, 0, -1};
+	Vec3f rayDir = Mat3x4f::rotationRad(sensorRotRad) * Vec3f{0, 0, 1};
```
Comment (PiotrMrozik) [inline test/src/scene/incidentAngleTest.cpp:41]: Those two lines are not very elegant. 
I get it, but those 1 and -1, are not very generic.

File: test/src/scene/incidentAngleTest.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	EXPECT_RGL_SUCCESS(rgl_graph_node_add_child(useRaysNode, transformRays));
+	EXPECT_RGL_SUCCESS(rgl_graph_node_add_child(transformRays, raytraceNode));
+	EXPECT_RGL_SUCCESS(rgl_graph_node_add_child(raytraceNode, yieldNode));
+
+	// Run graph
+	EXPECT_RGL_SUCCESS(rgl_graph_run(raytraceNode));
+
+	// Check results
+	TestPointCloud pc = TestPointCloud::createFromNode(yieldNode, yieldFields);
+	EXPECT_EQ(pc.getPointCount(), rays.size()); // All rays are expected to hit the wall
```
Comment (PiotrMrozik) [inline test/src/scene/incidentAngleTest.cpp:69]: Would change that comment to sth like:
No rays should be discarded.

File: src/gpu/optixPrograms.cu [hunk 3]
```diff
@@ -183,6 +190,16 @@ extern "C" __global__ void __closesthit__()
 		hitWorld = undistortedOrigin + undistortedDir * distance;
 	}
 
+	// Normal vector and incident angle
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/gpu/optixPrograms.cu:193]: Added to internal tasks, I will take it as soon as other high-priority tasks are finished.

File: src/gpu/optixPrograms.cu [hunk 4]
```diff
@@ -183,6 +190,16 @@ extern "C" __global__ void __closesthit__()
...
+	Vec3f rayDir = (hitWorld - origin).normalized();
+	const Vec3f wA = optixTransformPointFromObjectToWorldSpace(A);
+	const Vec3f wB = optixTransformPointFromObjectToWorldSpace(B);
+	const Vec3f wC = optixTransformPointFromObjectToWorldSpace(C);
+	const Vec3f wAB = wB - wA;
+	const Vec3f wBC = wC - wB;
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/gpu/optixPrograms.cu:199]: Good point, fixed.

File: test/include/helpers/geometryData.hpp [hunk 5]
```diff
@@ -37,19 +37,20 @@ static rgl_vec2f cubeUVs[] = {
...
-    {2, 5, 6},
-    {5, 4, 6},
-    {6, 4, 7},
-    {4, 0, 7},
-    {7, 0, 3},
-    {3, 2, 7},
-    {7, 2, 6},
-    {4, 5, 0},
-    {0, 5, 1}
+    {0, 3, 1},
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/include/helpers/geometryData.hpp:42]: Added to internal tasks.

File: test/src/scene/incidentAngleTest.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#include <helpers/sceneHelpers.hpp>
+#include <helpers/commonHelpers.hpp>
+#include <helpers/textureHelpers.hpp>
+#include <helpers/testPointCloud.hpp>
+
+#include <RGLFields.hpp>
+
+/**
+ * This test instantiates a large thin wall (100x100x0.1) at Z=10 and shoots parallel rays from different sensor positions and orientations.
+ * The wall is expected to be hit by all rays, and the incident angle is expected to be combination of sensor and box orientations.
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/scene/incidentAngleTest.cpp:12]: It's a leftover, good catch!

File: test/src/scene/incidentAngleTest.cpp (hunk 1 above)
Comment (prybicki) [reply test/src/scene/incidentAngleTest.cpp:41]: Added a comment.

File: test/src/scene/incidentAngleTest.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	for (int i = 0; i < pc.getPointCount(); ++i) {
+		EXPECT_TRUE(pc.getFieldValue<IS_HIT_I32>(i)); // All rays are expected to hit the wall
+
+		// Normal as expected
+		EXPECT_NEAR(pc.getFieldValue<NORMAL_VEC3_F32>(i).x(), expectedNormal.x(), EPSILON_F);
+		EXPECT_NEAR(pc.getFieldValue<NORMAL_VEC3_F32>(i).y(), expectedNormal.y(), EPSILON_F);
+		EXPECT_NEAR(pc.getFieldValue<NORMAL_VEC3_F32>(i).z(), expectedNormal.z(), EPSILON_F);
+
+		// Incident angle as expected
+		EXPECT_NEAR(pc.getFieldValue<INCIDENT_ANGLE_F32>(i), expectedIncidentAngleRad, 4 * EPSILON_F);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/scene/incidentAngleTest.cpp:79]: I think Copilot did a great job here:
image

**Review by PiotrMrozik — APPROVED** [review]

**Pull request merged**